src/sympc/protocol/falcon/falcon.py

```diff
@@ -73,93 +77,295 @@ def __eq__(self, other: Any) -> bool:
         return True
 
     @staticmethod
-    def mul_master(x: MPCTensor, y: MPCTensor, session: Session) -> MPCTensor:
+    def mul_master(
+        x: MPCTensor,
+        y: MPCTensor,
+        session: Session,
+        op_str: str,
+        kwargs_: Dict[Any, Any],
+    ) -> MPCTensor:
         """Master method for multiplication.
 
-        Performs Falcon's mul implementation, gets and reshares mul results and distributes shares.
-        This needs to be improved in future, it relies on orchestrator being a trusted third party.
-        Falcon, requires parties to be able to communication between each other.
-
         Args:
             x (MPCTensor): Secret
             y (MPCTensor): Another secret
             session (Session): Session the tensors belong to
+            op_str (str): Operation string.
+            kwargs_ (dict): Kwargs for some operations like conv2d
 
         Returns:
-            shares (ReplicatedSharedTensor): Shares in terms of ReplicatedSharedTensor.
+            MPCTensor: Result of the operation.
 
         Raises:
             ValueError: Raised when number of parties are not three.
-            NotImplementedError: Raised when implementation not present
-
+            ValueError : Raised when invalid security_type is provided.
         """
         if len(session.parties) != 3:
             raise ValueError("Falcon requires 3 parties")
 
         result = None
 
         if session.protocol.security_type == "semi-honest":
-            args = []
-            for index in range(0, 3):
-                args.append(
-                    [
-                        x.share_ptrs[index],
-                        y.share_ptrs[index],
-                    ]
-                )
-            z_shares_ptrs = parallel_execution(
-                Falcon.compute_zvalue_and_add_mask, session.parties
-            )(args)
-
-            result = MPCTensor(shares=z_shares_ptrs, session=x.session)
-            result.shape = MPCTensor._get_shape("mul", x.shape, y.shape)  # for prrs
-            result = ABY3.truncate(result, session)
-
+            result = Falcon.mul_semi_honest(x, y, session, op_str, kwargs_)
+        elif session.protocol.security_type == "malicious":
+            result = Falcon.mul_malicious(x, y, session, op_str, kwargs_)
         else:
-            raise NotImplementedError(
-                f"mult operation not implemented for {session.protocol.security_type} setting"
+            raise ValueError("Invalid security_type for Falcon multiplication")
+
+        result = ABY3.truncate(result, session)
+
+        return result
+
+    @staticmethod
+    def mul_semi_honest(
+        x: MPCTensor,
+        y: MPCTensor,
+        session: Session,
+        op_str: str,
+        kwargs_: Dict[Any, Any],
+        truncate: bool = True,
+    ) -> MPCTensor:
+        """Falcon semihonest multiplication.
+
+        Performs Falcon's mul implementation, gets and reshares mul results and distributes shares.
```

Comment:
_Performs Falcon's mul implementation, gets and reshares mul results and distributes shares_ , Change it to _Performs Falcon's mul implementation, add masks and performs resharing_ 

The description I had put earlier was verbose